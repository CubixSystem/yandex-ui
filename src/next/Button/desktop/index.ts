export * from '../Button@desktop';
export * from '../_baseline/Button_baseline';
export * from '../_pin/Button_pin_brick-brick';
export * from '../_pin/Button_pin_brick-circle';
export * from '../_pin/Button_pin_brick-clear';
export * from '../_pin/Button_pin_brick-round';
export * from '../_pin/Button_pin_circle-brick';
export * from '../_pin/Button_pin_circle-circle';
export * from '../_pin/Button_pin_circle-clear';
export * from '../_pin/Button_pin_clear-brick';
export * from '../_pin/Button_pin_clear-circle';
export * from '../_pin/Button_pin_clear-clear';
export * from '../_pin/Button_pin_clear-round';
export * from '../_pin/Button_pin_round-brick';
export * from '../_pin/Button_pin_round-clear';
export * from '../_size/Button_size_l';
export * from '../_size/Button_size_m';
export * from '../_size/Button_size_s';
export * from '../_theme/Button_theme_action@desktop';
export * from '../_theme/Button_theme_clear@desktop';
export * from '../_theme/Button_theme_link';
export * from '../_theme/Button_theme_normal@desktop';
export * from '../_theme/Button_theme_pseudo@desktop';
export * from '../_theme/Button_theme_raised@desktop';
export * from '../_theme/Button_theme_websearch@desktop';
export * from '../_type/Button_type_link';
export * from '../_type/Button_type_submit';
export * from '../_view/Button_view_action@desktop';
export * from '../_view/Button_view_clear@desktop';
export * from '../_view/Button_view_default@desktop';
export * from '../_view/Button_view_link@desktop';
export * from '../_view/Button_view_pseudo@desktop';
export * from '../_view/Button_view_raised@desktop';
export * from '../_width/Button_width_auto';
export * from '../_width/Button_width_max';
