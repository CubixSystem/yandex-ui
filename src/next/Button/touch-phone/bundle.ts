export * from '../Button.bundle/touch-phone';
