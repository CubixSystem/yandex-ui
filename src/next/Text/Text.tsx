import React, { CSSProperties, Ref, useState, useEffect, useRef, WeakValidationMap, ValidationMap } from 'react';
import { Composition } from '@bem-react/core';
import { cn } from '@bem-react/classname';
import { canUseDOM } from '../lib/canUseDOM';
import { mergeAllRefs } from '../lib/mergeRefs';
import type { TextOverflowValue } from './_overflow/Text_overflow';

type ComponentType = React.ComponentType<{
    ref?: Ref<HTMLElement>;
    style?: CSSProperties;
    className?: string;
}>;

export type TextElementType = keyof React.ReactHTML | ComponentType;

/**
 * Упрощенный ElementType для обхода
 * "Expression produces a union type that is too complex to represent. ts(2590)"
 */
type ShortElementType = 'a' | 'td' | 'span' | ComponentType;

type CommonProps = {
    /**
     * Задает отображение переполненного текста
     */
    overflow?: TextOverflowValue;

    /**
     * Максимальное количество строк текста (работает в связке с overflow)
     * @default 1
     */
    maxLines?: number;

    /**
     * Задает типографику текста в компоненте.
     */
    typography?: unknown;
}

type Props<T extends TextElementType> = CommonProps & {
    ref?: T extends keyof JSX.IntrinsicElements
        ? (JSX.IntrinsicElements[T] extends React.ClassAttributes<infer R> ? Ref<R> : never)
        : (T extends React.ComponentType<{ ref?: infer R }> ? R : never);

    /**
     * Тип элемента для отображения как (строка или компонент).
     * @default "span"
     */
    as?: T;
};

export type TextProps<T extends TextElementType> =
    Props<T> & Omit<React.ComponentProps<T>, keyof Props<T>>;

export const cnText = cn('Text');

const getLineHeight = (textElement: HTMLElement) => {
    return parseFloat(getComputedStyle(textElement).lineHeight);
};

const isFade = (overflow: TextOverflowValue) => {
    return overflow === 'fade' || overflow === 'fade-horizontal';
};

const isEllipsis = (overflow: TextOverflowValue) => {
    return overflow === 'ellipsis';
};

export type Text<U = {}> = {
    <T extends TextElementType = 'span'>(props: TextProps<T> & U): React.ReactElement | null;
    propTypes?: WeakValidationMap<CommonProps & U>;
    contextTypes?: ValidationMap<any>;
    defaultProps?: Partial<CommonProps & U>;
    displayName?: string;
}

export type PropsOfText<C extends Text<any>, T extends TextElementType = 'span'> =
    C extends Text<infer U> ? TextProps<T> & U : never;

/**
 * Базовый примитив представления текстовых данных.
 * @param { TextProps } props
 */
export const Text = React.memo(<T extends ShortElementType = 'span'>(props: TextProps<T>) => {
    const {
        as: As = 'span',
        ref,
        className,
        overflow,
        maxLines = 1,
        style = {},
        typography,
        ...otherProps
    } = props;
    const [, forceUpdate] = useState({});
    const textElement = useRef<HTMLElement>(null);

    useEffect(() => {
        if (textElement && overflow && isFade(overflow) && canUseDOM()) {
            forceUpdate({});
        }
    }, [typography, overflow]);

    const newStyle = { ...style };
    if (overflow) {
        if (isEllipsis(overflow)) {
            newStyle.WebkitLineClamp = maxLines;
        } else if (isFade(overflow)) {
            newStyle.whiteSpace = maxLines === 1 ? 'nowrap' : undefined;
            if (maxLines > 1) {
                const lineHeight = textElement.current && getLineHeight(textElement.current);
                if (typeof lineHeight === 'number' && !isNaN(lineHeight)) {
                    newStyle.maxHeight = `${maxLines * lineHeight}px`;
                }
            }
        }
    }

    return (
        <As
            ref={mergeAllRefs(ref, textElement)}
            style={newStyle}
            className={cnText({}, [className])}
            {...otherProps}
        />
    );
}) as Text;
Text.displayName = cnText();

export function enhanceText<H extends Composition<any>, P = {}>(
    hoc: H, Comp?: Text<P>,
): Text<P & (H extends Composition<infer T> ? T : never)> {
    return hoc(Comp || Text) as any;
}
