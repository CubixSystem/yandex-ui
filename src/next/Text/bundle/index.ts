export * from '../Text.bundle/common';
