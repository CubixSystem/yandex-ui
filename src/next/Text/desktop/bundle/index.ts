export * from '../../Text.bundle/desktop';
