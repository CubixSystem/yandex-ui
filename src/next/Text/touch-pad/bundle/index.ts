export * from '../../Text.bundle/touch-pad';
