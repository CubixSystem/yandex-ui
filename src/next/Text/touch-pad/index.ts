export * from '../Text@touch-pad';
export * from '../_typography/Text_typography';
export * from '../_typography/Text_typography_body-long-l';
export * from '../_typography/Text_typography_body-long-m';
export * from '../_typography/Text_typography_body-long-s';
export * from '../_typography/Text_typography_body-long-xl';
export * from '../_typography/Text_typography_body-short-l';
export * from '../_typography/Text_typography_body-short-m';
export * from '../_typography/Text_typography_body-short-s';
export * from '../_typography/Text_typography_body-short-xl';
export * from '../_typography/Text_typography_caption-l';
export * from '../_typography/Text_typography_caption-m';
export * from '../_typography/Text_typography_caption-xl';
export * from '../_typography/Text_typography_control-l';
export * from '../_typography/Text_typography_control-m';
export * from '../_typography/Text_typography_control-s';
export * from '../_typography/Text_typography_control-xl';
export * from '../_typography/Text_typography_control-xs';
export * from '../_typography/Text_typography_control-xxl';
export * from '../_typography/Text_typography_control-xxs';
export * from '../_typography/Text_typography_display-l';
export * from '../_typography/Text_typography_display-m';
export * from '../_typography/Text_typography_display-s';
export * from '../_typography/Text_typography_display-xl';
export * from '../_typography/Text_typography_headline-l';
export * from '../_typography/Text_typography_headline-m';
export * from '../_typography/Text_typography_headline-s';
export * from '../_typography/Text_typography_headline-xl';
export * from '../_typography/Text_typography_headline-xs';
export * from '../_typography/Text_typography_overline-l';
export * from '../_typography/Text_typography_overline-m';
export * from '../_typography/Text_typography_overline-s';
export * from '../_typography/Text_typography_subheader-l';
export * from '../_typography/Text_typography_subheader-m';
export * from '../_typography/Text_typography_subheader-s';
export * from '../_typography/Text_typography_subheader-xl';
export * from '../_weight/Text_weight';
export * from '../_weight/Text_weight_bold';
export * from '../_weight/Text_weight_light';
export * from '../_weight/Text_weight_medium';
export * from '../_weight/Text_weight_regular';
export * from '../_color/Text_color';
export * from '../_overflow/Text_overflow';
export * from '../_align/Text_align';
