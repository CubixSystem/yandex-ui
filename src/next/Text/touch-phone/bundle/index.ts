export * from '../../Text.bundle/touch-phone';
