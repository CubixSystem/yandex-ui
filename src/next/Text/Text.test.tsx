import React from 'react';
import { mount, render } from 'enzyme';
import { Text as TextCommon } from './Text';
import { Text as TextDesktop, cnText, TextAlignValue, TextOverflowValue } from './Text.bundle/desktop';
import { Text as TextTouchPhone } from './Text.bundle/touch-phone';
import { Text as TextTouchPad } from './Text.bundle/touch-pad';
import { serverEnvironmentSetup } from '../internal/utils';

const platforms = [['desktop', TextDesktop], ['touch-pad', TextTouchPad], ['touch-phone', TextTouchPhone]];

describe('Text@common', () => {
    describe('common environment', () => {
        test('должен быть установлен корректный displayName', () => {
            expect(TextCommon.displayName).toBe(cnText());
        });
    });
});

describe.each<any>(platforms)('Text@%s', (_platform, Text: typeof TextDesktop) => {
    describe('server environment', () => {
        serverEnvironmentSetup();

        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Text className="mix" />)).toMatchSnapshot();
        });
    });

    describe('client environment', () => {
        test('должен вернуть полный шаблон компонента (snapshot)', () => {
            expect(render(<Text className="mix" />)).toMatchSnapshot();
        });

        test('должен изменять тэг элемента через as', () => {
            const tagName = 'p';
            const wrapper = mount(<Text as={tagName} />);
            expect(wrapper.find(tagName)).toHaveLength(1);
        });

        test('должен добавлять выравнивание через props', () => {
            (['start', 'end', 'center', 'justify'] as TextAlignValue[]).forEach((align) => {
                expect(mount(<Text align={align} />).find(`span.${cnText()}_align_${align}`)).toHaveLength(1);
            });
        });

        test('должен управлять переполнением через props', () => {
            (['ellipsis', 'fade', 'fade-horizontal'] as TextOverflowValue[]).forEach((overflow) => {
                expect(mount(<Text overflow={overflow} />).find(`span.${cnText()}_overflow_${overflow}`)).toHaveLength(1);
            });
        });

        test('должен управлять цветом через props', () => {
            const color = 'primary';
            const wrapper = mount(<Text color={color} />);
            expect(wrapper.find(`span.${cnText()}_color_${color}`)).toHaveLength(1);
        });
    });
});
