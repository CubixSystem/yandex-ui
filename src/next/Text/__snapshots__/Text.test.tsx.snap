// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text@desktop client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text Text_weight_regular Text_color_primary mix"
/>
`;

exports[`Text@desktop server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text Text_weight_regular Text_color_primary mix"
/>
`;

exports[`Text@touch-pad client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text Text_weight_regular Text_color_primary mix"
/>
`;

exports[`Text@touch-pad server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text Text_weight_regular Text_color_primary mix"
/>
`;

exports[`Text@touch-phone client environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text Text_weight_regular Text_color_primary mix"
/>
`;

exports[`Text@touch-phone server environment должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Text Text_weight_regular Text_color_primary mix"
/>
`;
